pkg/vsphere/tags/tags.go

```diff
@@ -0,0 +1,233 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tags
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"regexp"
+	"strings"
+
+	"github.com/Sirupsen/logrus"
+	"github.com/pkg/errors"
+)
+
+const (
+	TagURL = "/com/vmware/cis/tagging/tag"
+)
+
+type TagCreateSpec struct {
+	CreateSpec TagCreate `json:"create_spec"`
+}
+
+type TagCreate struct {
+	CategoryID  string `json:"category_id"`
+	Description string `json:"description"`
+	Name        string `json:"name"`
+}
+
+type Tag struct {
+	ID          string   `json:"id"`
+	Description string   `json:"description"`
+	Name        string   `json:"name"`
+	CategoryID  string   `json:"category_id"`
+	UsedBy      []string `json:"used_by"`
+}
+
+var Logger = logrus.New()
+
+func (c *RestClient) CreateTagIfNotExist(ctx context.Context, name string, description string, categoryID string) (*string, error) {
+	tagCreate := TagCreate{categoryID, description, name}
+	spec := TagCreateSpec{tagCreate}
+	id, err := c.CreateTag(ctx, &spec)
+	if err == nil {
+		return id, nil
+	}
+	Logger.Debugf("Created tag %s failed for %s", errors.WithStack(err))
+	// if already exists, query back
+	if strings.Contains(err.Error(), ErrAlreadyExists) {
+		tagObjs, err := c.GetTagByNameForCategory(ctx, name, categoryID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to query tag %s for category %s", name, categoryID)
+		}
+		if tagObjs != nil {
+			return &tagObjs[0].ID, nil
+		}
+
+		// should not happen
+		return nil, errors.New("Failed to create tag for it's existed, but could not query back. Please check system")
+	}
+
+	return nil, errors.Wrap(err, "failed to create tag")
+}
+
+func (c *RestClient) DeleteTagIfNoObjectAttached(ctx context.Context, id string) error {
+	objs, err := c.ListAttachedObjects(ctx, id)
+	if err != nil {
+		return errors.Wrap(err, "failed to delete tag")
+	}
+	if objs != nil && len(objs) > 0 {
+		Logger.Debugf("tag %s related objects is not empty, do not delete it.", id)
+		return nil
+	}
+	return c.DeleteTag(ctx, id)
+}
+
+func (c *RestClient) CreateTag(ctx context.Context, spec *TagCreateSpec) (*string, error) {
+	Logger.Debugf("Create Tag %v", spec)
+	stream, _, status, err := c.call(ctx, "POST", TagURL, spec, nil)
+
+	Logger.Debugf("Get status code: %d", status)
+	if status != http.StatusOK || err != nil {
+		Logger.Debugf("Create tag failed with status code: %d, error message: %s", status, errors.WithStack(err))
+		return nil, errors.Wrapf(err, "Status code: %d", status)
+	}
+
+	type RespValue struct {
+		Value string
+	}
+
+	var pID RespValue
+	if err := json.NewDecoder(stream).Decode(&pID); err != nil {
+		Logger.Debugf("Decode response body failed for: %s", errors.WithStack(err))
+		return nil, errors.Wrap(err, "create tag failed")
+	}
+	return &(pID.Value), nil
```

Comment:
cosmetic: don't think the parens are need here, are they?  See this is a few places, just thought I'd mention it.